core/include/drivers/i2c.h

```diff
@@ -0,0 +1,351 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2022, Microchip
+ */
+
+#ifndef __DRIVERS_I2C_H
+#define __DRIVERS_I2C_H
+
+#include <kernel/dt.h>
+#include <kernel/dt_driver.h>
+#include <libfdt.h>
+#include <tee_api_types.h>
+
+/**
+ * DEFINE_I2C_DEV_DRIVER - Declare an I2C driver
+ *
+ * @__name: I2C device driver name
+ * @__match_table: match table associated to the driver
+ * @__i2c_probe: I2C probe function with the following prototype:
+ *	TEE_Result (*probe)(struct i2c_dev *i2c_dev, const void *fdt,
+ *			    int node, const void *compat_data);
+ */
+#define DEFINE_I2C_DEV_DRIVER(__name, __match_table, __i2c_probe) \
+	static TEE_Result __name ## _probe_i2c_dev(const void *fdt, int node, \
+						   const void *compat_data) \
+	{ \
+		struct i2c_dev *i2c_dev = NULL; \
+		TEE_Result res = TEE_ERROR_GENERIC; \
+\
+		res = i2c_dt_get_dev(fdt, node, &i2c_dev); \
+		if (res) \
+			return res; \
+		return __i2c_probe(i2c_dev, fdt, node, compat_data); \
+	} \
+	DEFINE_DT_DRIVER(__name ## _dt_driver) = { \
+		.name = # __name, \
+		.type = DT_DRIVER_I2C, \
+		.match_table = __match_table, \
+		.probe = __name ## _probe_i2c_dev, \
+	}
+
+#define I2C_SMBUS_MAX_BUF_SIZE	32
+
+enum i2c_smbus_dir {
+	I2C_SMBUS_READ,
+	I2C_SMBUS_WRITE,
+};
+
+enum i2c_smbus_protocol {
+	I2C_SMBUS_PROTO_BYTE,
+	/*
+	 * Like block but does not insert "count" in sent data, useful for
+	 * EEPROM read for instance which is not SMBus but requires such
+	 * sequence.
+	 */
+	I2C_SMBUS_PROTO_BLOCK_RAW,
+};
+
+struct i2c_ctrl;
+
+/**
+ * struct i2c_dev - I2C device
+ *
+ * @ctrl: I2C controller associated to the device
+ * @addr: device address on the I2C bus
+ * @priv: Private data associated to the device
+ */
+struct i2c_dev {
+	struct i2c_ctrl *ctrl;
+	uint8_t addr;
```

Comment:
Addresses can be 16bit. Maybe a later patch.

Reply:
I meant device I2C address can be 10bit.
Let's deal with that as a future evolution. A `uint16_t` here is not enough, we need some logic too.

Reply:
Yep, agreed, taht makes sense, I'll check if it is used in any other places.